_posts/2024-02-20-Navigating-Dependency-Chaos-with-Lockfiles.md

```diff
@@ -0,0 +1,348 @@
+
+```mermaid!
+flowchart LR
+
+    subgraph contextB["Context A"]
+        direction TB
+        processed1["data"]
+        widget
+        metadata1["metadata"]
+    end
+
+    subgraph contextA["Context B"]
+        direction TB
+        processed2["data"]
+        gizmo
+        metadata2["metadata"]
+    end
+
+    processed1 <-.-> | looosely coupled\nshared context | processed2
+    widget ~~~ | distinct solutions\nwith different names | gizmo
+    metadata1 ~~~ | distinct solutions\nwith the same names | metadata2
+```
+
+_Bounded contexts along with shared or distinct components can be used to help limit the complexity of a project in helpful ways._
+{:.center}
+
+The concept of [bounded context](https://martinfowler.com/bliki/BoundedContext.html) from [domain-driven design](https://en.wikipedia.org/wiki/Domain-driven_design) can sometimes be used to help distinguish what is in or out of scope for a particular project as a way of reducing complexity.
+Bounded context can be used as a way to draw abstract lines around a certain [span of control](https://en.wikipedia.org/wiki/Span_of_control) in order to align available resources (like time and people) with the focus of the project.
+It also can help promote [loose coupling](https://en.wikipedia.org/wiki/Loose_coupling) of software components in order to enable flexible design over time.
+Without these considerations and the use of dependencies we might face "endless" software [feature creep](https://en.wikipedia.org/wiki/Feature_creep) by continually adding new bounded contexts that are outside of our span of control (or resources).
+
+## Version constraints as dependency specification control
+
+<table>
+<tr>
+<th>Version constraint</th>
+<th>Description of the version constraint</th>
+</tr>
+<tr>
+<td>
+
+`==2.1.0`
+
+</td>
+<td>Exactly and only version 2.1.0</td>
+</tr>
+<tr>
+<td>
+
+`>=2.0.0`
+
+</td>
+<td>Greater than or equal to version 2.0.0</td>
+</tr>
+<tr>
+<td>
+
+`>=2.0.0, <3.0.0`
+
+</td>
+<td>Greater than or equal to version 2.0.0 and less than 3.0.0</td>
+</tr>
+<tr>
+<td>
+
+`>=2.0.0, <3.0.0, !=2.5.1`
+
+</td>
+<td>Greater than or equal to version 2.0.0, less than 3.0.0, and anything that's not exactly version 2.5.1</td>
+</tr>
+</table>
+
+_Version constraint specifications provide code-based descriptions for dependency versions within your project (Pythonic version specification examples above)._
+{:.center}
+
+Many aspects of dependency chaos arise from the fact that dependencies are updated at various times.
+We often want to make certain we use the most up-to-date version of a dependency because those updates may come with performance, corrective, or other benefits.
+To accomplish this we can use what are sometimes called dependency "version range constraints" or "compliant version specifications" to provide some flexibility in how packages are installed for our projects.
```

Comment:
I think it's worth mentioning that we have loose dependency specifications not just to incorporate bug fixes and new features, but because ultimately your software is going to have only a single version of each dependency installed (well, usually...) that's decided by all your other dependencies that specify that original dependency. If the version specs are too tight, then the transitive dependencies won't be resolvable as whole and you'd end up not being able to resolve dependencies for your library.